File: app/src/main/java/com/udacity/popularmovies/database/AppRepository.java
package com.udacity.popularmovies.database;

import android.content.Context;
import android.util.Log;

import androidx.lifecycle.LiveData;

import com.udacity.popularmovies.DetailsActivity;

import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class AppRepository {
    private static  AppRepository ourInstance;

    public LiveData<List<MovieEntity>> mMovies;
    private AppDatabase mDb;
    private static final String TAG = "AppRepository";

/*    All room database operations must be executed in a background thread
    One executor use for all db so that it has a queue if there are multiple db request*/
    private Executor executor = Executors.newSingleThreadExecutor();


    public static AppRepository getInstance(Context context) {
        if (ourInstance == null) {
            ourInstance = new AppRepository(context);
        }
        return ourInstance;
    }

    private AppRepository(Context context) {
        mDb = AppDatabase.getInstance(context);
        mMovies = getAllMovies();
        Log.d(TAG, "AppRepository: local data " + mMovies.getValue());
    }


    // Get data from DB
    private LiveData<List<MovieEntity>> getAllMovies(){
        return mDb.movieDao().getAll();
    }

    public void addMovieData() {

        final MovieEntity movie = DetailsActivity.getMovieDetails();

        executor.execute(new Runnable() {
            @Override
            public void run() {
                mDb.movieDao().insertMovie(movie);


            }
        });
    }

    public MovieEntity getMovieById(int movieId) {
        return mDb.movieDao().getMovieById(movieId);

    }

    public void deleteMovie(final MovieEntity movie) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                mDb.movieDao().deleteMovie(movie);
            }
        });
    }
}

File: app/src/main/java/com/udacity/popularmovies/database/AppRepository.kt
package com.udacity.popularmovies.database

import android.content.Context
import android.util.Log
import androidx.lifecycle.LiveData
import com.udacity.popularmovies.DetailsActivity
import java.util.concurrent.Executor
import java.util.concurrent.Executors

class AppRepository private constructor(context: Context) {
    var mMovies: LiveData<List<MovieEntity>>
    private val mDb: AppDatabase

    /*    All room database operations must be executed in a background thread
    One executor use for all db so that it has a queue if there are multiple db request*/
    private val executor: Executor = Executors.newSingleThreadExecutor()

    // Get data from DB
    private val allMovies: LiveData<List<MovieEntity>>
        private get() = mDb.movieDao().all

    fun addMovieData() {
        val movie: MovieEntity = DetailsActivity.movieDetails
        executor.execute { mDb.movieDao().insertMovie(movie) }
    }

    fun getMovieById(movieId: Int): MovieEntity? {
        return mDb.movieDao().getMovieById(movieId)
    }

    fun deleteMovie(movie: MovieEntity?) {
        executor.execute { mDb.movieDao().deleteMovie(movie) }
    }

    companion object {
        private var ourInstance: AppRepository? = null
        private const val TAG = "AppRepository"
        @JvmStatic
        fun getInstance(context: Context): AppRepository? {
            if (ourInstance == null) {
                ourInstance = AppRepository(context)
            }
            return ourInstance
        }
    }

    init {
        mDb = AppDatabase.getInstance(context)
        mMovies = allMovies
        Log.d(TAG, "AppRepository: local data " + mMovies.value)
    }
}